File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-0.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int n = s.size();
        unordered_map<char,int> mp;
        int i = 0, j = 0;
        int ans = 0;
        while (j < n) {
            mp[s[j]]++;
            while(mp[s[j]] > 1 && i <= j) {
                mp[s[i]]--;
                if(mp[s[i]] == 0) mp.erase(mp[s[i]]);
                i++;
            }
            ans = max(ans,j-i+1);
            j++;
        }

        return ans;
    }
};

File: Data Structures & Algorithms/split-array-largest-sum/submission-0.py
class Solution:
    def is_valid(self, nums:List[int], mid: int, k: int) -> bool:
        count = 1
        curr_sum = 0

        for num in nums:
            curr_sum += num
            if curr_sum > mid:
                count += 1
                curr_sum = num
                if count > k:
                    return False
        
        return True

    def splitArray(self, nums: List[int], k: int) -> int:
        low = max(nums)
        high = 0
        for num in nums:
            high += num
        
        ans = low

        while low <= high:
            mid = low + (high - low)//2
            if self.is_valid(nums,mid,k):
                ans = mid;
                high = mid - 1
            else:
                low = mid + 1
            
        return ans